src/main/java/com/minecolonies/coremod/entity/ai/citizen/cook/EntityAIWorkCook.java

```diff
@@ -0,0 +1,561 @@
+                    new InvWrapper((TileEntityChest) entity),
+                    predicate,
+                    Constants.STACKSIZE,
+                    new InvWrapper(worker.getInventoryCitizen()));
+        }
+        else if (entity instanceof TileEntityRack)
+        {
+            return InventoryUtils.transferXOfFirstSlotInProviderWithIntoNextFreeSlotInItemHandler(
+                    ((TileEntityRack) entity).getInventory(),
+                    predicate,
+                    Constants.STACKSIZE,
+                    new InvWrapper(worker.getInventoryCitizen()));
+        }
+        return false;
+    }
+
+    private AIState gatherFoodFromWarehouse()
+    {
+        if (!worker.getColony().hasWarehouse())
+        {
+            chatSpamFilter.requestTextComponentWithoutSpam(new TextComponentTranslation("com.minecolonies.coremod.ai.noWarehouse"));
+            return START_WORKING;
+        }
+
+        if (wareHouse == null)
+        {
+            double distance = Double.MAX_VALUE;
+            for (final AbstractBuilding building : worker.getColony().getBuildings().values())
+            {
+                if (building instanceof BuildingWareHouse && getOwnBuilding().getLocation().distanceSq(building.getLocation()) < distance)
+                {
+                    wareHouse = (BuildingWareHouse) building;
+                    distance = getOwnBuilding().getLocation().distanceSq(building.getLocation());
+                }
+            }
+        }
+
+        if (wareHouse == null)
+        {
+            Log.getLogger().warn("Colony should have warehouse but its not in the list of buildings!");
+            return START_WORKING;
+        }
+
+        if (walkTo == null)
+        {
+            if (walkToBlock(wareHouse.getLocation()))
+            {
+                return getState();
+            }
+
+            ((BuildingCook) getOwnBuilding()).setGatheredToday();
+            final BlockPos pos = wareHouse.getTileEntity().getPositionOfChestWithItemStack(isFood);
+            if (pos == null)
+            {
+                if(!((BuildingCook)getOwnBuilding()).hasGatheredToday())
+                {
+                    chatSpamFilter.talkWithoutSpam("com.minecolonies.coremod.ai.noFood");
+                }
+                return START_WORKING;
+            }
+            walkTo = pos;
+        }
+
+        if (walkToBlock(walkTo))
+        {
+            return getState();
+        }
+
+        int transfersDone = 0;
+        boolean transfered = true;
+        while (transfered && transfersDone < getOwnBuilding().getBuildingLevel())
+        {
+            transfered = tryTransferFromPosToCook(walkTo, isFood);
+            transfersDone++;
+        }
+
+        walkTo = null;
+        incrementActionsDone();
+        return START_WORKING;
+    }
+
+    @Override
+    protected int getActionsDoneUntilDumping()
+    {
+        return 1;
+    }
+
+    private AIState startWorking()
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this method 8, down to the maximum allowed 6. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "startWorking" is 20 which is greater than 15 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 25 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)